### PR TITLE
Fix tensor const correct

### LINKED ISSUE #467: fk::matrix<P> const and fk::vector<P> return mutable pointer.
**Describe the bug**

fk::matrix<P> const and fk::vector<P> return P* which can be modified. 

https://github.com/project-asgard/asgard/blob/410ca628d5d88da3989bac261f0edf9afde0176c/src/tensors.hpp#L408
https://github.com/project-asgard/asgard/blob/410ca628d5d88da3989bac261f0edf9afde0176c/src/tensors.hpp#L408

**To Reproduce**
1. git commit hash being built
2. cmake command
3. full program/test invocation command
4. additional steps 

A_copy is declared const [here](https://github.com/project-asgard/asgard/blob/develop/src/fast_math_tests.cpp#L802) but is modified by fm::gesv and therefore different on [line 810](https://github.com/project-asgard/asgard/blob/develop/src/fast_math_tests.cpp#L810)

**Expected behavior**

Same behavior as std::vector::data()
https://en.cppreference.com/w/cpp/container/vector/data

**System:**
 - system name [e.g. fusiont5, summit]
 - modules loaded [e.g. output of `module list`]
 - other systems where this is reproducible [e.g. "my laptop", "none"]

**Additional context**

### PR DESCRIPTION
**Pull request by mkstoyanov**
Please review the [developer documentation](https://github.com/project-asgard/asgard/wiki/developing)
on the wiki of this project that contains help and requirements.

## Proposed changes

Closes #467

Makes the tensors const-correct by properly overloading the `.data()` method.

Removed unnecessary code: `batch_gemv` and the raw-pointers were not used.

## What type(s) of changes does this code introduce?
_Put an `x` in the boxes that apply._

- [x] Bugfix
- [ ] New feature
- [x] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

### Does this introduce a breaking change?

- [x] Yes
- [ ] No

## What systems has this change been tested on?

Ubuntu 22.04

## Checklist

_Put an x in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- [x] this PR is up to date with current the current state of 'develop'
- [x] code added or changed in the PR has been clang-formatted
- [x] this PR adds tests to cover any new code, or to catch a bug that is being fixed
- [x] documentation has been added (if appropriate)

### CODE REVIEW

**Comment by mkstoyanov** [comment]
@quantumsteve finally we can fix this one, it should make your life a bit easier when fixing lib-dispatch and it will feed well into some of the ideas that I have of  improving the creation/deletion of transients.

File: src/lib_dispatch.cpp [hunk 1]
```diff
@@ -1052,13 +951,17 @@ void getrs(char *trans, int *n, int *nrhs, P *A, int *lda, int *ipiv, P *b,
   expect(*ldb >= 1);
   expect(*lda >= 1);
   expect(*n >= 0);
+
+  // the const_cast below is needed due to bad header under OSX
```
Comment (quantumsteve) [inline src/lib_dispatch.cpp:955]: We might try defining `ACCELERATE_NEW_LAPACK` in another PR and see what happens on MacOS.

https://developer.apple.com/documentation/accelerate/blas

**Review by quantumsteve — APPROVED** [review]
:tada:

File: src/lib_dispatch.cpp (hunk 1 above)
Comment (mkstoyanov) [reply src/lib_dispatch.cpp:955]: Maybe worth a try, but it reads more like they just have 64 bit indexing, which can be a problem if we pass in the sizes as pointers (compiler will automatically convert if we pass by value)

**Pull request merged**